### PR TITLE
Semantic Model Validation

### LINKED ISSUE #7969: [CT-2752] Run DSI SemanticManifest validations during parsing
In [dbt-semantic-interfaces](https://github.com/dbt-labs/dbt-semantic-interfaces/) there lives a tool called [SemanticManifestValidator](https://github.com/dbt-labs/dbt-semantic-interfaces/blob/adf7a992a231f5be6329e18b98c53f1d5b53db2a/dbt_semantic_interfaces/validations/semantic_manifest_validator.py#L57). The `SemanticManifestValidator` takes in a semantic manifest and runs a series of rules against it to ensure that it is valid. These are things like ensuring that either a `measure` has an `agg_time_dimension` set or that there is a `default_agg_time_dimension` set on the `semantic_model` that the `measure` belongs to.

Currently we don't run these validation rules in core. The only way someone using MetricFlow with core can run the validations is by doing
```bash
$ dbt parse
...
$ mf validate-configs
```
The issue with this is that validations are only run after the semantic manifest is written. This is problematic because validation issues should cause a failure before writing the semantic manifest. Additionally most people likely won't run `mf validate-configs` separately. Which means the issues would only be caught at query time 😬 We want to be informing customers of their issues when writing their configs, thus we need to ensure the `SemanticManifestValidator` is run during parsing.

Luckily, using the `SemanticManifestValidator` is fairly easy. It's a generic python class that requires that you declare an implementation of the `SemanticManifest` protocol on instantiation. It would look something like
```python
validator = SemanticManifestValidator[SemanticManifestImplementation]()
validation_results = validator.validate(semantic_manifest)
```

The other part of this is presently we only construct the `SemanticManifest`, and specifically a `PydanticSemanticManifest` right before we write out the semantic manifest. We can either reuse the [PydanticSemanticManifest builder](https://github.com/dbt-labs/dbt-core/blob/7156cc5c1da9da2dae9d7f2fd4d524e39773bb61/core/dbt/contracts/graph/manifest.py#L988C18-L988C19) to get a `PydanticSemanticManifest` and run that through a `SemanticManifestValidator` or we'll need to actually create a core implementation of the [SemanticManifest protocol](https://github.com/dbt-labs/dbt-semantic-interfaces/blob/main/dbt_semantic_interfaces/protocols/semantic_manifest.py) and then run that through a `SemanticManifestValidator`. The former is easier, but the latter is probably more in core's style.

# Acceptance Criteria

1. If a manifest has semantic models and/or metrics, the `SemanticManifestValidator` should get run as part of parsing
    - It needs happen after nodes have been fully "processed" (i.e. after things [like this](https://github.com/dbt-labs/dbt-core/blob/7156cc5c1da9da2dae9d7f2fd4d524e39773bb61/core/dbt/parser/manifest.py#L530))
    - It needs to happen before writing out the semantic manifest
3. Validation issues returned by the validation process should get logged
4. If there are `ERROR` level validation issues, parsing should be considered as having failed.

### PR DESCRIPTION
**Pull request by peterallenwebb**
resolves #7969

### Description

Support validation of Semantic Models by issuing a new log message type on validation warnings and errors, and throwing an exception at parse time when there are errors.

### Checklist

- [x] I have read [the contributing guide](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md) and understand what's expected of me
- [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
- [x] I have run this code in development and it appears to resolve the stated issue
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] I have [opened an issue to add/update docs](https://github.com/dbt-labs/docs.getdbt.com/issues/new/choose), or docs changes are not required/relevant for this PR
- [x] I have run `changie new` to [create a changelog entry](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md#adding-a-changelog-entry)

### CODE REVIEW

File: core/dbt/events/types_pb2.py [hunk 1]
```diff
@@ -16,3464 +15,10 @@
...
-_RUNRESULTWARNINGMESSAGEMSG = DESCRIPTOR.message_types_by_name['RunResultWarningMessageMsg']
-_DEBUGCMDOUT = DESCRIPTOR.message_types_by_name['DebugCmdOut']
-_DEBUGCMDOUTMSG = DESCRIPTOR.message_types_by_name['DebugCmdOutMsg']
-_DEBUGCMDRESULT = DESCRIPTOR.message_types_by_name['DebugCmdResult']
-_DEBUGCMDRESULTMSG = DESCRIPTOR.message_types_by_name['DebugCmdResultMsg']
-_LISTCMDOUT = DESCRIPTOR.message_types_by_name['ListCmdOut']
-_LISTCMDOUTMSG = DESCRIPTOR.message_types_by_name['ListCmdOutMsg']
-_NOTE = DESCRIPTOR.message_types_by_name['Note']
-_NOTEMSG = DESCRIPTOR.message_types_by_name['NoteMsg']
-EventInfo = _reflection.GeneratedProtocolMessageType('EventInfo', (_message.Message,), {
```
Comment (QMalcolm) [inline core/dbt/events/types_pb2.py:454]: It's weird that the next 3k lines were removed and not regenerated in any form 🤔

**Review by QMalcolm — COMMENTED** [review]
This looks great! Thanks again for doing this work.

As a side note, do you have any idea what's going on in `types_pb2.py`? This file gets auto generated, which it looks like you did. But it removed (and didn't regenerate) a whole section which seems odd 🤔

**Comment by QMalcolm** [comment]
We just merged #8050, pulling that in should fix the code quality check failure

File: core/dbt/events/types_pb2.py (hunk 1 above)
Comment (peterallenwebb) [reply core/dbt/events/types_pb2.py:454]: I looked into this a bit. Although my changeset removes all these lines, they were not present in older versions of the file. It looks like they may have been introduced unintentionally by [this changeset](https://github.com/dbt-labs/dbt-core/commit/4c44c29ee4cec511f4751cc9fec668949ff9726b).

**Comment by peterallenwebb** [comment]
@QMalcolm I'd argue this is ready to merge. I disabled the check we added while pairing because several existing tests have metrics without semantic models, and I'd like to move ahead with this and fix that separately. I'm happy to follow that up in this sprint if we have time.

**Comment by QMalcolm** [comment]
> @QMalcolm I'd argue this is ready to merge. I disabled the check we added while pairing because several existing tests have metrics without semantic models, and I'd like to move ahead with this and fix that separately. I'm happy to follow that up in this sprint if we have time.

Yea in the `depends_on` work I'm doing I'm fixing those tests 😂 I'll rebase after this is merged, and then uncomment the check 👍

**Review by QMalcolm — APPROVED** [review]
I'm still confused about why we're generated so much less code in `types_pb2`, but everything seems happy so 🤷 Looks good!

**Pull request merged**